fcs/match.py

```diff
@@ -71,43 +116,185 @@ def get_all_non_candidates(domain, vat=None):
     ).first()
     if not u or not u.oldcompany_verified:
         return None
+    link = (
+        models.OldCompanyLink.query.filter_by(
+            undertaking_id=undertaking_id,
+            oldcompany_id=u.oldcompany_id).first()
+    )
+    if link:
+        link.verified = False
+        link.date_verified = None
 
     u.oldcompany_verified = False
     u.oldcompany_account = None
     u.oldcompany_extid = None
     u.oldcompany = None
-    log_match(undertaking_id, False, user)
+    log_match(undertaking_id, None, False, user)
     models.db.session.commit()
     return u
 
 
 def verify_none(undertaking_id, domain, user):
+    return verify_manual(undertaking_id, domain, None, user)
+
+
+def verify_manual(undertaking_id, domain, oldcompany_accout, user):
     u = models.Undertaking.query.filter_by(
         external_id=undertaking_id,
         domain=domain
     ).first()
-
     if not u:
         return None
     if u.oldcompany_verified is True:
         return u
     u.oldcompany = None
     u.oldcompany_verified = True
-    u.oldcompany_account = None
+    u.oldcompany_account = oldcompany_accout
     u.oldcompany_extid = None
-    if call_bdr(u, old_collection=None):
-        log_match(undertaking_id, True, user,
-                  oldcompany_account=None)
+    if call_bdr(u, old_collection=oldcompany_accout):
+        log_match(undertaking_id, None, True, user,
+                  oldcompany_account=oldcompany_accout)
         models.db.session.commit()
         send_match_mail(match=False, user=user, company_name=u.name,
                         company_id=u.external_id)
     else:
         models.db.session.rollback()
     return u
+
+
+def get_country(country_code):
+    cc = country_code.lower()
+    if cc == 'gb':
+        cc = 'uk'
+    elif cc == 'el':
+        cc = 'gr'
+    return cc
+
+
+def has_match(company, old):
+    c_code = get_country(company['country_code'])
+    o_code = get_country(old['country_code'])
+    if c_code != o_code:
+        return False
+    c_vat = company['vat'] or ''
+    o_vat = old['vat_number'] or ''
+    if all((c_vat, o_vat)) and c_vat == o_vat:
+        return True
+    c_name = company['name'].lower()
+    o_name = old['name'].lower()
+    return str_matches(c_name, o_name)
+
+
+def match_all(companies, oldcompanies):
+    links = []
+    new_companies = []
+    companies = [c.as_dict() for c in companies]
+    oldcompanies = [o.as_dict() for o in oldcompanies]
+
+    for c in companies:
+        candidates = [o for o in oldcompanies if has_match(c, o)]
+        if candidates:
+            links.append((c, candidates))
+        else:
+            new_companies.append(c)
+    for company, candidates in links:
+        for old in candidates:
+            add_link(company['id'], old['id'])
+    models.db.session.commit()
+    return links, new_companies
+
+
+@match_manager.command
+def run():
+    companies = get_unverified_companies()
+    oldcompanies = get_oldcompanies_for_matching()
```

Comment:
Add an exception here: if AUTO_VERIFY_ALL_COMPANIES, dict like {domain: True/False}. If domain is true, mark company as verified from this stage.